### PR TITLE
Documentation: Changing "combinatorial" to "combinational"

### LINKED ISSUE #1301: Question: Combinational vs Combinatorial
The [language guide](https://amaranth-lang.org/docs/amaranth/latest/guide.html) makes heavy reference to "combinatorial" logic.

In his reference text, _Digital Design: Principles and Practice_, Wakerly writes:

> A step _backward_ in MMI's introduction of PAL devices was their popularization of the word "combinatorial" to describe combinational circuits. _Combinational_ circuits have no memory -- their output at any time depends on the current input _combination_. For well-rounded computer engineers, the word "combinatorial" should conjure up vivid images of binomial coefficients, problem-solving complexity, and computer-science-great Donald Knuth.

[1, p. 343]

On the other hand, the Wikipedia article, [Combinational Logic](https://en.wikipedia.org/wiki/Combinational_logic), indicates that it is sometimes referred to as "combinatorial logic" and cites Maxfield's _FPGAs: World Class Designs_:

> Some folks prefer to say "combinational logic," while others favor "combinatorial logic."
> **1822:** England. Charles Babbage starts to build a mechanical calculating machine called the Difference Engine.

[[2, p. 16](https://books.google.com/books?id=kQuOKBSOz5QC&dq=%22combinatorial+logic%22&pg=PA16#v=onepage&q=%22combinatorial%20logic%22&f=false)]

So, perhaps the convention is not so clear cut.
Is Wakerly just being a curmudgeon? Is Maxfield actually citing something that Charles Babbage said? It isn't clear.

In any case, if you were in favor of changing the documentation to say "combinational" rather than "combinatorial", I would be happy to submit a pull-request for same.

---

[1] John F. Wakerly, _Digital Design: Principles and Practice_ (3rd Edition). Upper Saddle River, New Jersey, USA: Prentice hall, 2001.

[2] Clive Maxfield, _FPGAs: World Class Designs_. Newnes, 2009.

### PR DESCRIPTION
**Pull request by X-Illuminati**
Resolves #1301

The first two commits update the documentation and are simple find-and-replace in python docstrings and rst, respectively.
When rebasing, an additional search may be needed.

The third commit adds additional content and should be reviewed by project maintainers more closely.

### CODE REVIEW

**Review by whitequark — CHANGES_REQUESTED** [review]
- Please rebase on top of https://github.com/amaranth-lang/amaranth/pull/1333 once it's in (I completely forgot that it was on a branch and not merged).
- Please move your style guide sentence to the section added in #1333.
- Please squash all commits.
- It's not necessary to add a compatibility anchor.

**Comment by X-Illuminati** [comment]
@whitequark , ready for review

**Comment by X-Illuminati** [comment]
@whitequark , heh you are too quick for me. Will take a look at #1333 and rebase.

File: docs/contrib.rst [hunk 1]
```diff
@@ -136,6 +136,9 @@ Some of the formatting guidelines are:
 * For properties, phrase the short description (first line of docstring) like ``Value of thing.``, i.e. as a declarative sentence.
 * When documenting signatures of interfaces, as well as components, use the (non-standard) `Members` section to document their interface members, and only that section; do not document them in an `Attributes` section.
 * If an anchor for a section is needed, namespace it, e.g. the ``.. _lang-assignable:`` anchor is a part of the ``lang`` namespace. Anchor names are global.
+* It is preferred to use the term "combinational" rather than "combinatorial" to refer to non-sequential logic. See `Issue #1301`_ for more details.
```
Comment (whitequark) [inline docs/contrib.rst:139]:
```suggestion
* To refer to non-sequential logic, use the term "combinational" over "combinatorial".
```

(A style guide's purpose is to set the standard. It is stronger than a mere _preference_--you are required to use this term when contributing--and it's also not particularly important what the reasoning is, just like for all other items here.)

**Comment by X-Illuminati** [comment]
@whitequark rebased and ready for review

**Review by whitequark — APPROVED** [review]
Thanks!

**Pull request merged**